codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/SerializeMiddleware.java

```diff
@@ -0,0 +1,165 @@
+                            return next.HandleSerialize(ctx, in)
+                        }
+
+                        """, MapUtils.of(
+                        "context", SmithyGoDependency.CONTEXT.interfaceSymbol("Context"),
+                        "input", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeInput"),
+                        "handler", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeHandler"),
+                        "output", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeOutput"),
+                        "metadata", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("Metadata"),
+                        "opName", serialName,
+                        "body", generateHandleSerializeBody())
+        );
+    }
+
+    private GoWriter.Writable generateHandleSerializeBody() {
+        return goTemplate("""
+                    $route:W
+
+                    $serialize:W
+
+                """,
+                MapUtils.of(
+                        "route", handleHttp(),
+                        "serialize", handlePayload()
+                ));
+    }
+
+    private GoWriter.Writable handleHttp() {
+        String contentType = "application/x-amz-json-1.0";
+
+        return goTemplate("""
+                    input, ok := in.Parameters.($input:P)
+                    if !ok {
+                        return out, metadata, $errorf:T("unexpected input type %T", in.Parameters)
+                    }
+                    _ = input
+
+                    req, ok := in.Request.($request:P)
+                    if !ok {
+                        return out, metadata, $errorf:T("unexpected transport type %T", in.Request)
+                    }
+
+                    req.Method = $methodPost:T
+                    req.URL.Path = "/"
+                    req.Header.Set("Content-Type", $contentType:S)
+                    req.Header.Set("X-Amz-Target", $target:S)
+
+                """,
+        MapUtils.of(
+                "input", ctx.getSymbolProvider().toSymbol(input),
+                "request", SMITHY_HTTP_TRANSPORT.pointableSymbol("Request"),
+                "methodPost", GoStdlibTypes.Net.Http.MethodPost,
+                "contentType", contentType,
+                "target", ctx.getService().getId().getName() + '.' + operation.getId().getName(),
+                "errorf", GoStdlibTypes.Fmt.Errorf
+                ));
+    }
+
+    private GoWriter.Writable handlePayload() {
+        return goTemplate("""
+
+                jsonEncoder := $encoder:T()
+
+                err = $serialize:L(input, jsonEncoder.Value)
+                if err != nil {
+                    return out, metadata, &$error:T{Err: err}
+                }
+
+                payload := $reader:T(jsonEncoder.Bytes())
+                req, err = req.SetStream(payload)
+                if err != nil {
+                    return out, metadata, &$error:T{Err: err}
+                }
+
+                in.Request = req
+            """,
+    MapUtils.of(
```

Comment:
JFYI, don't need to address it now, but pay attention to your spacing/indentation as you work. The indentation here is off.